quickwit-config/src/source_config.rs

```diff
@@ -0,0 +1,174 @@
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::path::{Path, PathBuf};
+
+use quickwit_common::uri::Uri;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct SourceConfig {
+    pub source_id: String,
+    #[serde(flatten)]
+    pub source_type: SourceType,
+}
+
+impl SourceConfig {
+    pub fn source_type(&self) -> &str {
+        match self.source_type {
+            SourceType::File(_) => "file",
+            SourceType::Kafka(_) => "kafka",
+            SourceType::Kinesis(_) => "kinesis",
+            SourceType::Vec(_) => "vec",
+            SourceType::Void(_) => "void",
+        }
+    }
+
+    // TODO: Remove after source factory refactor.
+    pub fn params(&self) -> serde_json::Value {
+        match &self.source_type {
+            SourceType::File(params) => serde_json::to_value(params),
+            SourceType::Kafka(params) => serde_json::to_value(params),
+            SourceType::Kinesis(params) => serde_json::to_value(params),
+            SourceType::Vec(params) => serde_json::to_value(params),
+            SourceType::Void(params) => serde_json::to_value(params),
+        }
+        .unwrap()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "source_type", content = "params")]
+pub enum SourceType {
+    #[serde(rename = "file")]
+    File(FileSourceParams),
+    #[serde(rename = "kafka")]
+    Kafka(KafkaSourceParams),
+    #[doc(hidden)]
+    #[serde(rename = "kinesis")]
+    Kinesis(KinesisSourceParams),
+    #[serde(rename = "vec")]
+    Vec(VecSourceParams),
+    #[serde(rename = "void")]
+    Void(VoidSourceParams),
+}
+
+impl SourceType {
+    pub fn file<P: AsRef<Path>>(filepath: P) -> Self {
+        Self::File(FileSourceParams::file(filepath))
+    }
+
+    pub fn stdin() -> Self {
+        Self::File(FileSourceParams::stdin())
+    }
+
+    pub fn void() -> Self {
+        Self::Void(VoidSourceParams)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct FileSourceParams {
+    /// Path of the file to read. Assume stdin if None.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[serde(deserialize_with = "absolute_filepath_from_str")]
+    pub filepath: Option<PathBuf>, //< If None read from stdin.
+}
+
+// Deserializing a filepath string into an absolute filepath.
+fn absolute_filepath_from_str<'de, D>(deserializer: D) -> Result<Option<PathBuf>, D::Error>
```

Comment:
here it is canonical.
absolute is slightly different from canonical.
canonical resolve links.

It's raises an interesting question. Here we might actually prefer absolute.